pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -1339,6 +1396,118 @@ func (c *controller) UpdateNodeTerminationCondition(ctx context.Context, machine
+	// Log the error message for machine failure
+	klog.Error(description)
+
+	clone.Status.LastOperation = v1alpha1.LastOperation{
+		Description:    description,
+		State:          v1alpha1.MachineStateFailed,
+		Type:           machine.Status.LastOperation.Type,
+		LastUpdateTime: metav1.Now(),
+	}
+	clone.Status.CurrentStatus = v1alpha1.CurrentStatus{
+		Phase: v1alpha1.MachineFailed,
+		//TimeoutActive:  false,
+		LastUpdateTime: metav1.Now(),
+	}
+
+	_, err := c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(ctx, clone, metav1.UpdateOptions{})
+	updated := false
+	if err != nil {
+		// Keep retrying until update goes through
+		klog.Errorf("update failed for machine %q in function. Retrying, error: %q", machine.Name, err)
+	} else {
+		updated = true
+		klog.Infof("Machine State has been updated for %q with providerID %q and backing node %q", machine.Name, getProviderID(machine), getNodeName(machine))
+		// Return error for continuing in next iteration
+		err = fmt.Errorf("machine creation is successful. Machine State has been UPDATED")
+	}
+
+	return machineutils.ShortRetry, updated, err
+}
+
+func (c *controller) canMarkMachineFailed(machineSetName, machineDeployName, namespace string, maxReplacements int) (bool, error) {
+	var (
+		list     = []string{machineDeployName}
+		selector = labels.NewSelector()
+		req, _   = labels.NewRequirement("name", selection.Equals, list)
+	)
+
+	selector = selector.Add(*req)
+	//listing all machines for machinedeployment
+	machineList, err := c.machineLister.Machines(namespace).List(selector)
+	if err != nil {
+		return false, err
+	}
+
+	klog.Infof("num machines in machineDeploy=%s is %d", machineDeployName, len(machineList))
+	//inProgress keeps count of number of machines which are counted as `getting replaced`
+	var inProgress int
+
+	var terminating, failed, pending, noPhase, crashLooping int
+
+	for _, machine := range machineList {
+		if machine.Status.CurrentStatus.Phase != v1alpha1.MachineUnknown && machine.Status.CurrentStatus.Phase != v1alpha1.MachineRunning {
+			inProgress++
+			// `Terminating` is not considered as replaced because during replacement old machines are always in terminating
+			// state and so can't allow them to consume maxReplacement count
+			if machine.Status.CurrentStatus.Phase != v1alpha1.MachineTerminating {
+
+				if machine.Status.CurrentStatus.Phase == v1alpha1.MachineFailed {
+					failed++
+				} else if machine.Status.CurrentStatus.Phase == v1alpha1.MachinePending {
+					pending++
+				} else if machine.Status.CurrentStatus.Phase == "" {
+					noPhase++
+				} else if machine.Status.CurrentStatus.Phase == v1alpha1.MachineCrashLoopBackOff {
+					crashLooping++
+				}
+			} else {
+				terminating++
+				// counting terminated machine twice in `inProgress` to avoid case where there is delay by MS controller
+				// in adding new machine object and terminating machines are also gone.
+				inProgress++
+			}
```

Comment:
switch/case would improve readability

Reply:
Yes these counters are for better logging. I would use switch case for the same.

Reply:
addressed in 2aa4416f08be317fc00c9b12355fe15ea986b5a8